text/0089-scoped-api-keys.md

```diff
@@ -0,0 +1,128 @@
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the main key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the client code. |
+| Scoped API Key     | These keys are not stored and managed by a MeiliSearch instance. They are generated for each end-user by the backend code from a MeiliSearch API Key. They are used by the end-users to only search the documents belonging to them. |
+| Multi-Tenancy      | By multi-tenancy, we mean that an end-user only accesses data belonging to him within an index shared with other end-users. |
+
+### IV. Personas
+
+| Persona | Role |
+|---------|------|
+| Mark    | Mark is a developer for a SaaS company. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
+| UserX   | UserX represents any end-user searching from the frontend interfaces provided by Anna and Mark's company. |
+
+### V. `Scoped API Key` Explanations
+
+#### Summary Key Points
+
+- `Scoped API keys` are generated from a MeiliSearch parent `API key` on the client's server code to resolve multi-tenancy use-case by restricting access to data within an index according to the criteria chosen by the team managing a MeiliSearch instance.
+- These `Scoped API keys` can't be less restrictive than a parent `API key` and can only be used for the search action with a predefined forced filter field.
+- These `Scoped API keys` are not stored and thus retrievable on MeiliSearch. This is why we highly advise setting an expiration time on that type of API key for security reasons.
+
+#### Solving Multi-Tenancy with `Scoped API Keys`
+
+![](https://i.imgur.com/J4jVe1n.png)
+
+Let's say that `Mark` is a developer for a SaaS platform. He would like to ensure that an end-user can only access his documents at search time. **His database contains many users and he hopes to have many more in the future.**
+
+When a user registers, the backend-side client code generates a `Scoped API Key` specifically for that end-user so he can only access his documents.
+
+A `filter` parameter is set to restrict the search for documents having a `user_id` attribute. This `filter` parameter is contained in the `Scoped API key` payload and cannot be modified during the search by the end-user making the request. `filter` can be made of any valid filters. e.g. `user_id = 10 and category = Romantic`
+
+This `Scoped API Key` is generated from a parent `API key` used to cipher the `Scoped API Key`. On the MeiliSearch side, this permits checking permissions and authorizing the `UserX` search request using this `Scoped API Key`.
+
+---
+
+#### Generating a `Scoped API Key`
+
+```javascript
+const scopedApiKeyRestrictions = {
+     "indexesPolicy": {
+         "products": {
+             "filter": "user_id = 1"
+         },
+         "reviews": {
+             "filter": "user_id = 1 AND published = true"
```

Comment:
This example is a bit confusing as `products` do not have an associated `user_id` normally. Since we are not a relational database I'm not sure what use cases it represents

Reply:
Imagine that you have 2 indexes: `products` and `reviews` and you only want end-users to retrieve documents containing `user_id = :userId`. You also want to ensure that end-users will only access published reviews when searching documents from the `reviews` index.